firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -403,34 +394,28 @@ private void insertOrUpdatePrefs(PersistedInstallationEntry prefs) {
    */
   private PersistedInstallationEntry getPrefsWithGeneratedIdMultiProcessSafe() {
     synchronized (lockGenerateFid) {
-      CrossProcessLock lock =
-          CrossProcessLock.acquire(firebaseApp.getApplicationContext(), LOCKFILE_NAME_GENERATE_FID);
-      try {
-        PersistedInstallationEntry prefs =
-            persistedInstallation.readPersistedInstallationEntryValue();
-        // Check if a new FID needs to be created
-        if (prefs.isNotGenerated()) {
-          // For a default firebase installation read the existing iid. For other custom firebase
-          // installations create a new fid
-
-          // Only one single thread from one single process can execute this block
-          // at any given time.
-          String fid = readExistingIidOrCreateFid(prefs);
-          prefs =
-              persistedInstallation.insertOrUpdatePersistedInstallationEntry(
-                  prefs.withUnregisteredFid(fid));
-        }
-        return prefs;
-
-      } finally {
-        // It is possible that the lock acquisition failed, resulting in lock being null.
-        // We handle this case by going on with our business even if the acquisition failed
-        // but we need to be sure to only release if we got a lock.
-        if (lock != null) {
-          lock.releaseAndClose();
-        }
-      }
+      return CrossProcessLock.whileLocked(
+          firebaseApp.getApplicationContext(),
+          LOCKFILE_NAME_GENERATE_FID,
+          this::readLatestPersistedInstallationsMultiProcessSafe);
+    }
+  }
+
+  private PersistedInstallationEntry readLatestPersistedInstallationsMultiProcessSafe() {
+    PersistedInstallationEntry prefs = persistedInstallation.readPersistedInstallationEntryValue();
```

Comment:
It's fine for this PR, but this is a dependency that violates Single Responsibility Principle.
A class called `CrossProcessLock` should not know anything about persisted installations.
We should segregate these responsibilities, have one class that does nothing but get a CrossProcssLock and/or executes methods synchronized.
Another class should know methods like reading and writing to disk.

The business logic should then just call
{pseudo-code}
`CrossProcessLock.supplySynchronized(PersistedInstallation::readFromDisc)`